### PR TITLE
[SCB-1803] kie support long polling

### PR DESCRIPTION
**Pull request by GuoYL123**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java [hunk 1]
```diff
@@ -48,16 +49,27 @@
...
+    thread.setName("org.apache.servicecomb.config.kie");
+    thread.setDaemon(true);
+    return thread;
+  });
 
   private static final long TIME_OUT = 10000;
 
+  private static AtomicBoolean IS_FIRST_PULL = new AtomicBoolean(true);
+
+  private static final int LONG_POLLING_WAIT_TIME = 30;
```
Comment (liubao68) [inline dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: long polling timeout is equals to client request timeout, will this cause client timeout frequently?

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java [hunk 2]
```diff
@@ -114,17 +129,22 @@ public void run() {
...
     }
 
-    //todo : latch down
     @SuppressWarnings("deprecation")
     void refreshConfig() {
       String path = "/v1/"
           + KieConfig.INSTANCE.getDomainName()
           + "/kie/kv?label=app:"
           + KieConfig.INSTANCE.getAppName();
+      if (!IS_FIRST_PULL.get()) {
```
Comment (liubao68) [inline dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: Maybe need check enableLongPolling?

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java [hunk 3]
```diff
@@ -140,23 +160,23 @@ void refreshConfig() {
...
               } else if (rsp.statusCode() == HttpStatus.SC_NOT_FOUND) {
                 EventManager.post(new ConnSuccEvent());
-//                latch.countDown();
               } else {
                 EventManager.post(new ConnFailEvent("fetch config fail"));
                 LOGGER.error("Config update from {} failed. Error message is [{}].",
                     serviceUri,
                     rsp.statusMessage());
               }
+              if (enableLongPolling) {
```
Comment (liubao68) [inline dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: Maybe need check if first call?

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java (hunk 1 above)
Comment (GuoYL123) [reply dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: After checking , vert.x default strategy is never timeout. I set the idle timeout 50s.

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java (hunk 2 above)
Comment (GuoYL123) [reply dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: done

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java (hunk 3 above)
Comment (GuoYL123) [reply dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: first time also need this code

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java [hunk 4]
```diff
@@ -84,6 +101,8 @@ private void deployConfigClient() throws InterruptedException {
     Vertx vertx = VertxUtils.getOrCreateVertxByName("kie", vertxOptions);
 
     HttpClientOptions httpClientOptions = new HttpClientOptions();
+    httpClientOptions.setKeepAlive(true);
+    httpClientOptions.setIdleTimeout(SOCKET_TIMOUT);
```
Comment (liubao68) [inline dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: Please look at RestTransportClient to setting HttpClientOptions, if keepAliveTimeout not set, IdelTimeout may not work as expected. SOCKET_TIMEOUT should better be IdleTimeout and maybe better add a configuration for this.

**Comment by liubao68** [comment]
I added a sub task for documentation: https://issues.apache.org/jira/browse/SCB-1807

File: dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java (hunk 4 above)
Comment (GuoYL123) [reply dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: keepAliveTimeout default is 60s

Comment (GuoYL123) [reply dynamic-config/config-kie/src/main/java/org/apache/servicecomb/config/kie/client/KieClient.java]: I see the vert.x said :

>    * Set the keep alive timeout for HTTP/1.x, in seconds.
>    * This value determines how long a connection remains unused in the pool before being evicted and closed. 

I think the default value is ok

**Pull request merged**